### ISSUE TITLE
Request::withUri does not respect the $preserveHost argument as per PSR7

### ISSUE THREAD

**Issue by jleeothon** [guzzle/psr7#223]
From the PSR7 specification:

```
     * You can opt-in to preserving the original state of the Host header by
     * setting `$preserveHost` to `true`. When `$preserveHost` is set to
     * `true`, this method interacts with the Host header in the following ways:
     *
     * - If the Host header is missing or empty, and the new URI contains
     *   a host component, this method MUST update the Host header in the returned
     *   request.
     * - If the Host header is missing or empty, and the new URI does not contain a
     *   host component, this method MUST NOT update the Host header in the returned
     *   request.
     * - If a Host header is present and non-empty, this method MUST NOT update
     *   the Host header in the returned request.
```

tl;dr the Host header may be updated even if the`$preserveHost` parameter is `true`.

However, the implementation as per v1.4.2 does nothing to the Host header when `$preserveHost` is `true`. See: https://github.com/guzzle/psr7/blob/1.4.2/src/Request.php#L113

**Comment by gmponos** [comment]
Is this issue for 1.5 also?

**Comment by Tobion** [comment]
yes
